component-api/src/main/java/org/talend/sdk/component/api/record/OrderedMap.java

```diff
@@ -0,0 +1,310 @@
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package org.talend.sdk.component.api.record;
+
+import java.util.Collections;
+import java.util.HashMap;
+import java.util.Iterator;
+import java.util.Map;
+import java.util.NoSuchElementException;
+import java.util.Optional;
+import java.util.function.Consumer;
+import java.util.function.Function;
+import java.util.stream.Stream;
+import java.util.stream.StreamSupport;
+
+import lombok.AllArgsConstructor;
+import lombok.Getter;
+
+/**
+ * On TCK Record, we have to control order of elements with keeping efficient access.
+ * LinkedHashMap has efficient access, but order of element is only the order of insertions.
+ * List (ArrayList, LinkedList) allow to fine control order but have inefficent access.
+ * This class aims to control element order with keeping efficient access.
+ * 
+ * @param <T> : type of element.
+ */
+public class OrderedMap<T> {
+
+    @AllArgsConstructor
+    static class Node<T> implements Iterable<Node<T>> {
+
+        @Getter
+        public T entry;
+
+        public Node<T> next;
+
+        public Node<T> prec;
+
+        public Node<T> insert(final T newEntry) {
+            final Node<T> newNode = new Node<>(newEntry, this.next, this);
+            return this.insert(newNode);
+        }
+
+        public Node<T> insert(final Node<T> newNode) {
+            if (newNode == this) {
+                return newNode;
+            }
+            if (this.next != null) {
+                this.next.prec = newNode;
+            }
+            newNode.prec = this;
+            newNode.next = this.next;
+            this.next = newNode;
+
+            return newNode;
+        }
+
+        public void remove() {
+            if (next != null) {
+                this.next.prec = this.prec;
+            }
+            if (this.prec != null) {
+                this.prec.next = this.next;
+            }
+            this.next = null;
+            this.prec = null;
+        }
+
+        @Override
+        public Iterator<Node<T>> iterator() {
+            return new EntryNodeIterator(this);
+        }
+
+        @AllArgsConstructor
+        static class EntryNodeIterator<T> implements Iterator<Node<T>> {
+
+            private Node<T> current;
+
+            @Override
+            public boolean hasNext() {
+                return current != null;
+            }
+
+            @Override
+            public Node<T> next() {
+                if (!this.hasNext()) {
+                    throw new NoSuchElementException("no further node");
+                }
+                final Node<T> next = this.current;
+                this.current = next.next;
+                return next;
+            }
+        }
+    }
+
+    private Node<T> first;
+
+    private Node<T> last;
+
+    private final Map<String, Node<T>> entryIndex;
+
+    private final Function<T, String> getName;
```

Comment:
Nitpicking: `getNameSupplier`? `getNameFunction`?

Reply:
I thought here that "get" would already imply that it's a function ?

Reply:
`getNamingSupplier`, `getIdentitySupplier` ?

Reply:
👍